First release

jannatkhandev opened the pull request:
A fully functional end-to-end test RC app for Todoist integration. 
_closes_ https://github.com/RocketChat/Rocket.Chat/issues/34896

🚀 Looking forward to add many more features.


https://github.com/user-attachments/assets/fd1f88e9-05f8-430a-9fb1-9819d4a896b7


Todoist App provides you the following slash commands, /todoist:

  1. **help**: shows this list.
  2. **auth**: starts the process to authorize your Todoist Account.
  3. **task**: opens modal to create a new task.
  4. **projects**: fetches the projects you are part of.
  5. **tasks**: fetches the active tasks.
  6. **sections**: fetches the active tasks.
  7. **labels**: fetches your personal labels.
  8. **shared-labels**: fetches your shared labels.


Allows users to create tasks from message, create task within project.

jannatkhandev committed:
First release

jannatkhandev commented:
@casalsgh sorry to tag you, but this is ready to merge.
Please take a look.
Best,
Jannat.

Gustrb commented:
Hey, @jannatkhandev just to make you aware, this is a 4k+ line pull request, the review will take more time than it would be since, I can't look at this PR in my free time. So I'll have to check if it makes sense with the rest of the company for me to help you through it.
But don't worry, I didn't forgot about you :)

Gustrb commented on README.md:
```diff
@@ -1,2 +1,60 @@
+The application allows users to create and manage thier projects and tasks, create new tasks, edit and delete their tasks, sections, comments & labels and do much more right from Rocket.Chat.
+
+
+<h2>🚀 Features </h2>
```
not sure why you are mixing markdown and html in the readme

Gustrb commented on README.md:
```diff
@@ -1,2 +1,60 @@
+
+
+<h2>🔧 Installation steps </h2>
+
```
```suggestion
1. Install Rocket.Chat apps CLI: https://developer.rocket.chat/docs/apps-engine-cli
```

Gustrb commented on README.md:
```diff
@@ -1,2 +1,60 @@
+	 `npm install`
+
+ 3. Deploy Rocket.Chat app: </br>
+    `rc-apps deploy --url http://localhost:3000 --username user_username --password user_password`
```
I guess you could reword this to be something like this:

```suggestion
    `$ rc-apps deploy --url <deployment_url> --username <user_username> --password <user_password>`
```

Gustrb commented on TodoistApp.ts:
```diff
@@ -0,0 +1,98 @@
+import { MiscEnum } from './src/enums/Misc';
+
+export class TodoistApp extends App {
+    public botUsername: string;
```
put the constructor after the properties

Gustrb commented on TodoistApp.ts:
```diff
@@ -0,0 +1,98 @@
+    };
+
+    private async authorizationCallback(token: IAuthData, user: IUser, read: IRead, modify: IModify, http: IHttp, persistence: IPersistence) {
+        if (token) {
```
what if there is no token? can this happen? shouldn't it be encoded in the type system?

```ts
private async authorizationCallback(token: IAuthData | undefined, ...)
```

Gustrb commented on TodoistApp.ts:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+    public async onEnable(): Promise<boolean> {
+        this.botUsername = 'todoist-app.bot';
```
why not have this defined in the class itself, no need to set this property in the onEnable method

Gustrb commented on TodoistApp.ts:
```diff
@@ -0,0 +1,98 @@
+    }
+    public async onEnable(): Promise<boolean> {
+        this.botUsername = 'todoist-app.bot';
+        this.botUser = (await this.getAccessors().reader.getUserReader().getByUsername(this.botUsername)) as IUser;
```
getUserReader can return undefined, you should check to see if this.botUser is not null, and maybe return false and log.
Something like this:
```ts
this.botUser = (await this.getAccessors().reader.getByUsername(this.botUsername));
if (!this.botUser) {
    this.getLogger().error('Failed to get the user for the bot, something, please ensure the user exists, re-install the app if not.');
    return false;
}
return true;
```

Gustrb replied:
Also, you can do this to get the appUser:
```ts
const appUser = (await read.getUserReader().getAppUser());
```

jannatkhandev replied:
> Also, you can do this to get the appUser:
> 
> ```ts
> const appUser = (await read.getUserReader().getAppUser());
> ```

getAppUser returns IUser | undefined, the former implementation works fine, wdyt?

jannatkhandev replied:
I removed onEnable as nevertheless we weren't referencing this.botUser rather using 
`const appUser = (await read.getUserReader().getAppUser()) as IUser;`

Gustrb commented on TodoistApp.ts:
```diff
@@ -0,0 +1,98 @@
+
+    public async executeBlockActionHandler(context: UIKitBlockInteractionContext, read: IRead, http: IHttp, persistence: IPersistence, modify: IModify): Promise<IUIKitResponse> {
+        const handler = new ExecuteBlockActionHandler(this, read, http, modify, persistence);
+        return await handler.run(this, context, read, http, persistence, modify);
```
no need to await here

Gustrb commented on TodoistApp.ts:
```diff
@@ -0,0 +1,98 @@
+        await sendDirectMessage(read, modify, user, text, persistence);
+    }
+
+    public getOauth2ClientInstance(): IOAuth2Client {
```
why not do this at the top class level(or in the constructor, there is no need to have a signleton there)
make it a simple getter
like:
```ts
private oauth2ClientInstance: IOAuth2Client = createOAuth2Client(this, this.oauth2Config);
private httpHelperInstance: HttpHelper = new HttpHelper(this, this.getAcessors().http);

```

Gustrb commented on src/commands/subcommands/authorize.ts:
```diff
@@ -0,0 +1,27 @@
+  const url = await app.getOauth2ClientInstance().getUserAuthorizationUrl(user);
+  const block: LayoutBlock[] = [];
+
+  let authButton = await getButton('Authorize', '', '', '', 'primary', url.toString());
```
I believe it would much clearer if you created the button object here, or called a method like:
`getAuthorizeButton` and pass only the URL

Gustrb replied:
Btw, this is not a variable, const it please

Gustrb commented on src/commands/subcommands/authorize.ts:
```diff
@@ -0,0 +1,27 @@
+  );
+  block.push(textsectionBlock);
+
+  await sendDirectMessage(read, modify, user, '', persistence, block);
```
why not just
```ts
await sendDirectMessage(read, modify, user, '', persistence, [textsectionBlock]);
```
no need for the block variable

Gustrb commented on src/commands/subcommands/authorize.ts:
```diff
@@ -0,0 +1,27 @@
+  const block: LayoutBlock[] = [];
+
+  let authButton = await getButton('Authorize', '', '', '', 'primary', url.toString());
+  let textsectionBlock = await getSectionBlock(
```
this is a constant

Gustrb commented on src/commands/subcommands/labels.ts:
```diff
@@ -0,0 +1,85 @@
+  const user = context.getSender();
+  const room = context.getRoom();
+
+  const response = await app.getHttpHelperInstance().get(user, getLabelsUrl());
```
please use guard clauses to make it easier to read
```ts
if (response.statusCode !== HttpStatusCode.Ok) {
    const msg = modify
      .getCreator()
      .startMessage()
      .setText(`❗️ Unable to retrieve labels! \n Error ${JSON.stringify(response)}`)
      .setRoom(room);
    await modify.getNotifier().notifyUser(user, msg.getMessage());
    return;
}

if (!response.data || response.data.length === 0) {
    const msg = modify
        .getCreator()
        .startMessage()
        .setText('No personal labels found. Create one using the Todoist app or website.')
        .setRoom(room);
      await modify.getNotifier().notifyUser(user, msg.getMessage());
      return;
}
```

Gustrb commented on src/commands/subcommands/labels.ts:
```diff
@@ -0,0 +1,85 @@
+    const block: LayoutBlock[] = [];
+
+    for (const label of response.data) {
+      let labelNameBlock = await getSectionBlock(`${label.name}`);
```
isn't label.name a string?

Gustrb commented on src/commands/subcommands/labels.ts:
```diff
@@ -0,0 +1,85 @@
+    const builder = modify.getCreator().startMessage().setRoom(room);
+    const block: LayoutBlock[] = [];
+
+    for (const label of response.data) {
```
I think you could do it different:
```ts
const blocks: LayoutBlock[] = (await Promise.all(response.data.map(this.createLabelSection))).flat();
builder.setBlocks(block);
await modify.getNotifier().notifyUser(user, builder.getMessage());

//...
private async createLabelSection(label: /* idk the type */) {
    const labelContextBlock = await getContextBlock(
        `Color: ${String(label.color).charAt(0).toUpperCase() + String(label.color).slice(1)} | ` +
          `Order: ${label.order} | Favorite: ${label.is_favorite ? 'Yes' : 'No'}`
    );

    const shareButton = await getButton(
        MiscEnum.SHARE_LABEL_BUTTON,
        MiscEnum.LABEL_ACTIONS_BLOCK,
        MiscEnum.SHARE_LABEL_ACTION_ID,
        `${label.id}`,
        'primary'
    );
    const deleteButton = await getButton(
        MiscEnum.DELETE_LABEL_BUTTON,
        MiscEnum.LABEL_ACTIONS_BLOCK,
        MiscEnum.DELETE_LABEL_ACTION_ID,
        `${label.id}`,
        'danger'
    );
    const actionBlock = await getActionsBlock(MiscEnum.LABEL_ACTIONS_BLOCK, [
        shareButton,
        deleteButton,
    ]);
    return [labelNameBlock, labelContextBlock, actionBlock];
}
```

Gustrb reviewed (changes requested):
I couldn't review everything, but since the code is mostly similar, I think you'll be able to extrapolate a lot of those comments to more places

jannatkhandev commented:
Thanks for the detailed feedback, I will ensure I inspect rest of the code as well.
I really appreciate your time.
Best,
Jannat.

jannatkhandev committed:
Address review

jannatkhandev committed:
Remove unused import and lint

jannatkhandev commented:
Hey there,
I have made changes as per the review, please take a look.

Gustrb commented on TodoistApp.ts:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public getHttpHelperInstance(): HttpHelper {
+        if (!this.httpHelperInstance) {
```
hey you should do the same thing as you did with the oauth2client.
since the property is not static each instance is going to have its own `HttpHelper`, so it doesn't make a lot of sense to follow the `Singleton` pattern here, just put it into the constructor, and turn this into a simple getter

jannatkhandev replied:
After implementing your suggestion, I'm getting an error `Cannot read properties of undefined (reading 'http')`. Could this be because accessors.http isn't available yet when the constructor runs? 
I notice that for oauth2ClientInstance we use a factory function createOAuth2Client() rather than direct instantiation. 
Should we follow a similar pattern for HttpHelper, or is there a different way to safely access the http accessor?

Gustrb replied:
Yea that might be the case. Can't you access htttp from inside the app class?

Gustrb replied:
this probably happens because .http at construction is undefined, so when you handle it to a function undefined is passed by value to it, so it does not have visibility when the value gets set.
If you can't get via the app class, don't worry about this one, it is a small thing :)

jannatkhandev replied:
I'm still getting the same 'Cannot read properties of undefined (reading 'http')' error even when using this.getAccessors().http. I tried both direct accessor access and getAccessors() in the constructor, but neither seems to work.
Shall I go back to lazy initialization in getHttpHelperInstance()

Gustrb replied:
Yea, go ahead :)

Gustrb commented on src/commands/subcommands/labels.ts:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  const builder = modify.getCreator().startMessage().setRoom(room);
+  const blocks = (await Promise.all(response.data.map(createLabelSection))).reduce(
```
can't you just flat this?
I guess it is simpler than doing a full-fletched functional fold onto it

jannatkhandev replied:
.flat throws error
```
  ›   Error:    ⟿  Error src/commands/subcommands/labels.ts (54,77): Property 'flat' does not exist on type 'any[]'. Do you 
 ›   need to change your target library? Try changing the 'lib' compiler option to 'es2019' or later.
```
as Array.prototype.flat() method was introduced in ES2019 (ES10)
Our code uses this method, but the [CLI's TypeScript environment](https://github.com/RocketChat/Rocket.Chat.Apps-cli/blob/master/tsconfig.json) doesn't recognize it because it's not included in its type definitions

Gustrb replied:
Oh, ok then, makes sense :)

Gustrb commented on src/handlers/viewSubmitInteraction.ts:
```diff
@@ -0,0 +1,66 @@
+  public async run(context: UIKitViewSubmitInteractionContext) {
+    const data = context.getInteractionData();
+    let { view, room } = data;
+    const elements = view.id.split('#');
```
I advise to checking viewId and roomId, I know they are fixed fields, but checking them gives us less mental overhead when we make mistakes.
```ts
...
if (!viewId || !roomId) {
  this.app.getLogger().warn('Invalid view id');
  // send error message
  return context.getInteractionResponder().errorResponse();
}
```

You can do something like this:
```ts
const [viewId, roomId] = view.id.split('#');

if (!room) {
  this.app.getLogger().warn('Room data not present in UI Kit context');
  room = await.this.read.getRoomReader().getById(roomId);
  if (!room) {
    this.app.getLogger().warn(`Room with id: ${roomId} does not exist`);
    // probably send a message with an error message so the user knows what happened.
   return context.getInteractionResponder().errorResponse();
  }
}
```

Gustrb commented on src/handlers/viewSubmitInteraction.ts:
```diff
@@ -0,0 +1,66 @@
+    if (!room) room = await this.read.getRoomReader().getById(roomId);
+    if (!room) return context.getInteractionResponder().errorResponse();
+    try {
+      switch (viewId) {
```
my advice, log what came from viewId and roomId, so you can visibility when things stop working

Gustrb commented on src/helpers/const.ts:
```diff
@@ -0,0 +1,80 @@
+  RemoveSharedLabel: 'labels/shared/remove',
+};
+
+export const getProjectsUrl = () => {
```
does this have to be a function?
```ts
const PROJECTS_URL = `${ApiBaseUrl}/${ApiVersion.V2}/${ApiEndpoint.Projects}`;
...
```

jannatkhandev replied:
Just for consistency with others that have param like:
```typescript
const PROJECT_URL = (projectId: string) => `${ApiBaseUrl}/${ApiVersion.V2}/${ApiEndpoint.Project(projectId)}`;
```

Gustrb commented on src/helpers/message.ts:
```diff
@@ -0,0 +1,125 @@
+    return;
+  }
+
+  if (room) {
```
let's rewrite this part to make it more clear:
```ts
if (room) {
  return room;
}

const newRoom = modify
  .getCreator()
  .startRoom()
  .setType(RoomType.DIRECT_MESSAGE)
  .setCreator(appUser)
  .setMembersToBeAddedByUsernames(usernames);
const roomId = await modify.getCreator().finish(newRoom);
return read.getRoomReader().getById(roomId);
```

Gustrb commented on src/helpers/message.ts:
```diff
@@ -0,0 +1,125 @@
+      .setType(RoomType.DIRECT_MESSAGE)
+      .setCreator(appUser)
+      .setMembersToBeAddedByUsernames(usernames);
+    roomId = await modify.getCreator().finish(newRoom);
```
also, I think this method can throw an exception, so you should probably try to make sure you return undefined if there was an error creating the room

Gustrb commented on src/helpers/message.ts:
```diff
@@ -0,0 +1,125 @@
+ * @returns the room or undefined if botUser or botUsername is not set
+ */
+export async function getDirect(
+  read: IRead,
```
please create an object with the params of this function, it reduces the mental overhead

Gustrb commented on src/helpers/message.ts:
```diff
@@ -0,0 +1,125 @@
+  }
+}
+
+export async function sendMessage(
```
here also, use an object for the parameters

Gustrb commented on src/helpers/message.ts:
```diff
@@ -0,0 +1,125 @@
+    .setParseUrls(false)
+    .setText(message);
+
+  if (blocks !== undefined) {
```
I guess you could just
```ts
if (blocks)
```

so you don't call setBlocks on an empty array, not that big of a deal though

Gustrb commented on src/helpers/message.ts:
```diff
@@ -0,0 +1,125 @@
+}
+
+export async function sendNotification(
+  read: IRead,
```
add an object to carry the params

Gustrb commented on src/helpers/message.ts:
```diff
@@ -0,0 +1,125 @@
+}
+
+export async function sendDirectMessage(
+  read: IRead,
```
also in here

Gustrb commented on src/helpers/message.ts:
```diff
@@ -0,0 +1,125 @@
+  persistence: IPersistence,
+  user: IUser
+): Promise<boolean> {
+  const notificationsController = new NotificationsController(read, persistence, user);
```
I don't think this is a good approach.
for every message you are creating a controller, instead of just passing the userId and read and persistence to a function.

Gustrb commented on src/helpers/notification.ts:
```diff
@@ -0,0 +1,65 @@
+export class NotificationsController {
+  private read: IRead;
+  private persistence: IPersistence;
+  private association: RocketChatAssociationRecord;
```
tise should not be a class member

Gustrb commented on src/lib/create/createTask.ts:
```diff
@@ -0,0 +1,66 @@
+      1
+  );
+
+  const body = {
```
you should do some error handling before calling the http server, this way you can avoid some bogus calls

Gustrb commented on src/lib/delete/deleteItem.ts:
```diff
@@ -0,0 +1,168 @@
+  }
+}
+
+export async function deleteTask(params: DeleteHandlerParams) {
```
I'm not sure this function makes a lot of sense

jannatkhandev replied:
Just helper fn for code clarity.

Gustrb commented on src/lib/share/shareComment.ts:
```diff
@@ -0,0 +1,15 @@
+import { getCommentUrl } from '../../helpers/const';
+import { shareItem } from './shareItem';
+
+export async function shareComment(params) {
```
why not just have a function to format, and leave it to the caller?
this seems an unnecessary wrapper

jannatkhandev replied:
Creating a function to format but leaving the wrapper for code clarity.

Gustrb commented on src/lib/share/shareItem.ts:
```diff
@@ -0,0 +1,48 @@
+  modify: IModify;
+  getUrl: (id?: string) => string;
+  formatMessage: (response: any) => string;
+  method?: HttpMethod; // Optional - defaults to GET
```
you don't have to mark it as optional in the type, the function signature already ensures that for you

jannatkhandev committed:
Address second review

jannatkhandev commented:
Thanks for the detailed feedback. I have addressed the review and have made changes accordingly.
Also have tested my app end to end.

jannatkhandev commented:
> Thanks for the detailed feedback. I have addressed the review and have made changes accordingly. Also have tested my app end to end.

I really appreciate your time, I shall be highly grateful to you :)

Gustrb commented on src/commands/subcommands/authorize.ts:
```diff
@@ -0,0 +1,39 @@
+    read: read,
+    modify: modify,
+    user: user,
+    message: '',
```
shouldn't this be the default?

Gustrb commented on src/commands/subcommands/labels.ts:
```diff
@@ -0,0 +1,81 @@
+    const msg = modify
+      .getCreator()
+      .startMessage()
+      .setText(`❗️ Unable to retrieve labels! \n Error ${JSON.stringify(response)}`)
```
This often leaks stuff to the user, avoid returning network responses, those belong to logs.
Besides that, the message is good :)

Gustrb commented on src/commands/subcommands/labels.ts:
```diff
@@ -0,0 +1,81 @@
+async function createLabelSection(label: ILabel): Promise<LayoutBlock[]> {
+  const labelNameBlock = getSectionBlock(label.name);
+  const labelContextBlock = getContextBlock(
+    `Color: ${String(label.color).charAt(0).toUpperCase() + String(label.color).slice(1)} | ` +
```
Lets put this into a separate function wdyt?
this function is dealing contextually with label blocks, I don't think it is good for it to know exactly how the color field is internally

Gustrb commented on src/commands/subcommands/labels.ts:
```diff
@@ -0,0 +1,81 @@
+  const room = context.getRoom();
+
+  const response = await app.getHttpHelperInstance().get(user, getLabelsUrl());
+
```
I have an issue with two things over here:
1. We don't know the type of label, which has some annoying issues such as having to look at the todoist docs to know what is being returned, we would benefit from having the type in here.
2. We don't validate anything, so if, for some reason, they do a breaking change, everything is going to stop working and we won't know why.

My proposal is the following. Instead of having this HttpHelper, create services for the entities, such as LabelService, which will have an API, such as: fetchLabels(user) and will return a list of labels, and will perform schema validation on the json response.

This will solve both of the problems and will also help us put network related stuff away to this more of UI-focused code. I think it would valuable, wdyt?

jannatkhandev replied:
Doing exactly that now, but keeping HttpHelper.

Gustrb commented on src/commands/subcommands/projects.ts:
```diff
@@ -0,0 +1,102 @@
+import { getProjectsUrl } from '../../helpers/const';
+import { IProject } from '../../interfaces/projects';
+
+export async function projects(
```
Hey please take a look at the suggestions I pointed out in regards to Labels, I think they are going to be usefull in here as well, since the code is really similar

Gustrb commented on src/commands/subcommands/sections.ts:
```diff
@@ -0,0 +1,81 @@
+import { getSectionsUrl } from '../../helpers/const';
+import { ISection } from '../../interfaces/sections';
+
+export async function sections(
```
Take a look at what I said about labels, I think it is going to be usefull here as well :)

Gustrb commented on src/commands/subcommands/sharedLabels.ts:
```diff
@@ -0,0 +1,90 @@
+  const user = context.getSender();
+  const room = context.getRoom();
+
+  const response = await app.getHttpHelperInstance().get(
```
Let's create a SharedLabelsService.
That has a method called getSharedLabelsOmittingPersonalLabels, so this part will be a lot cleaner

Gustrb commented on src/commands/subcommands/task.ts:
```diff
@@ -0,0 +1,19 @@
+
+  const modal = await createTaskModal({ projectId: undefined, roomId: room.id });
+
+  if (triggerId) {
```
looking nice, good work :)

Gustrb commented on src/commands/subcommands/tasks.ts:
```diff
@@ -0,0 +1,104 @@
+import { getTasksUrl } from '../../helpers/const';
+import { ITask } from '../../interfaces/tasks';
+
+export async function tasks(
```
I think we should have a service to talk with the network, refer back to what I pointed out in labels

Gustrb commented on src/enums/Misc.ts:
```diff
@@ -0,0 +1,61 @@
+export enum MiscEnum {
```
I think you could give better names to your enums.
App.ts, Misc.ts are really not well separated.
add as many enums as you need, there is no need to be cheap with it.

Gustrb commented on src/handlers/actionButtonInteraction.ts:
```diff
@@ -0,0 +1,54 @@
+import { MiscEnum } from '../enums/Misc';
+import { createTaskModal } from '../modals/createTaskModal';
+
+export class ExecuteActionButtonHandler {
```
Hey, rename the file to be capitalized and make it match to the class name

Gustrb commented on src/handlers/blockInteraction.ts:
```diff
@@ -0,0 +1,78 @@
+import { shareTask } from '../lib/share/shareTask';
+import { createTaskModal } from '../modals/createTaskModal';
+
+export class ExecuteBlockActionHandler {
```
hey, rename the file to be capitalized and to match the class name

Gustrb commented on src/handlers/viewSubmitInteraction.ts:
```diff
@@ -0,0 +1,76 @@
+    const data = context.getInteractionData();
+    let { view, room } = data;
+    const elements = view.id.split('#');
+    const viewId = elements[0];
```
why do you need `elements` variable? doesn't this do the trick?
```ts
const [viewId, roomId] = view.id.split('#');
```

Gustrb commented on src/handlers/viewSubmitInteraction.ts:
```diff
@@ -0,0 +1,76 @@
+      }
+    }
+
+    logger.debug(viewId, room.id);
```
I would add some better debuging info, this is not useful at all, if it all blows app you won't even find this

Gustrb commented on src/handlers/viewSubmitInteraction.ts:
```diff
@@ -0,0 +1,76 @@
+import { createTask } from '../lib/create/createTask';
+import { deleteComment, deleteLabel, deleteSection, deleteTask } from '../lib/delete/deleteItem';
+
+export class ExecuteViewSubmitHandler {
```
hey, rename the file to match the class name

Gustrb replied:
Also, I think it should be implementing an interface, it is just working due to duck typing

Gustrb commented on src/helpers/blockBuilder.ts:
```diff
@@ -0,0 +1,275 @@
+  | 'time_picker'
+  | 'toggle_switch';
+
+interface InputParams extends BaseElementParams {
```
Why not use tagged unions instead of inheritance here

jannatkhandev replied:
The inheritance pattern is deliberate here since blockId, actionId, and appId are fundamental properties shared across ALL input types, not mutually exclusive variations. A tagged union would suggest these base properties vary by type, which isn't the case. The current approach reduces duplication while maintaining type safety through the type discriminator and options object structure.

Gustrb commented on src/helpers/const.ts:
```diff
@@ -0,0 +1,80 @@
+const ApiBaseUrl: string = 'https://api.todoist.com/rest';
+
+const ApiVersion = {
+  V1: 'v1',
```
v1 is not used anywhere, it could be in the base url

Gustrb commented on src/helpers/http.ts:
```diff
@@ -0,0 +1,79 @@
+
+export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
+
+export interface IHttpOptions extends IHttpRequest {
```
I think we can remove this class in favor of service classes that wrap the IHttp from apps engine

Gustrb commented on src/lib/create/createTask.ts:
```diff
@@ -0,0 +1,78 @@
+      .startMessage()
+      .setText(`❗️ Unable to create task! \n Error: ${error}}`)
+      .setRoom(room!);
+    await modify.getNotifier().notifyUser(user, msg.getMessage());
```
you should return here, otherwise you're doing the call anyway

Gustrb commented on src/lib/create/createTask.ts:
```diff
@@ -0,0 +1,78 @@
+    ...(project_id && { project_id }),
+    due_date: taskdueDate
+      ? new Date(taskdueDate).toISOString().split('T')[0] // Formatted as YYYY-MM-DD
+      : new Date(Date.now() + 86400000).toISOString().split('T')[0], // Tomorrow's date
```
have tomorrow as a constant, then you won't need this comment

Gustrb commented on src/lib/create/createTask.ts:
```diff
@@ -0,0 +1,78 @@
+    content: `${taskName}`,
+    ...(description && { description }),
+    ...(project_id && { project_id }),
+    due_date: taskdueDate
```
this date formatting looks sketchy

jannatkhandev replied:
I have formatted it in the format as expected by todoist, any points on how should I do it better?

Gustrb commented on src/lib/delete/deleteItem.ts:
```diff
@@ -0,0 +1,168 @@
+  itemType,
+}: DeleteItemParams) {
+  const data = context.getInteractionData();
+  const itemId = data.view.submit!.value;
```
don't assert, check

Gustrb commented on src/lib/delete/deleteItem.ts:
```diff
@@ -0,0 +1,168 @@
+  const user = data.user;
+  const roomId = room.id;
+  let roomFromId = await read.getRoomReader().getById(roomId);
+  if (!room) room = roomFromId!;
```
Fetch the room from id only if it exists:
```ts
if (!room) {
    // log that now room was passed in the context
    room = await read.getRoomReader().getById(roomId);
    if (!room) {
        // error out and remember to log what happened
    }
}

Gustrb commented on src/lib/delete/deleteItem.ts:
```diff
@@ -0,0 +1,168 @@
+  let roomFromId = await read.getRoomReader().getById(roomId);
+  if (!room) room = roomFromId!;
+  try {
+    const response = await app.getHttpHelperInstance().delete(user, getUrl(itemId));
```
please wrap this in a service call so you don't have to handle with http codes inside a domain context

Gustrb commented on src/lib/delete/deleteItem.ts:
```diff
@@ -0,0 +1,168 @@
+    ...params,
+    getUrl: getTaskUrl,
+    itemType: 'Task',
+    app: params.app!,
```
https://www.allthingstypescript.dev/i/138198930/what-about-non-null-assertions

Gustrb commented on src/lib/delete/deleteItem.ts:
```diff
@@ -0,0 +1,168 @@
+  });
+}
+
+export async function deleteSection(params: DeleteHandlerParams) {
```
I think having a service will fix this weird pattern

Gustrb commented on src/lib/fetch/fetchComments.ts:
```diff
@@ -0,0 +1,105 @@
+  };
+}
+
+export async function getComments(
```
Put the call part inside a service and formatting can be here.
Also, I don't understand why the folders are called like that.
Separate stuff by domain not by name.
i.g. instead of having a folder called share
have a folder called comment and add comment stuff inside there

Gustrb commented on src/lib/fetch/fetchComments.ts:
```diff
@@ -0,0 +1,105 @@
+  const blockId = data.blockId;
+
+  const isProject = blockId.includes('project');
+  const url = getCommentsUrl(isProject ? undefined : itemId, isProject ? itemId : undefined);
```
defer this to a service class

Gustrb commented on src/lib/share/shareItem.ts:
```diff
@@ -0,0 +1,48 @@
+  try {
+    const response = await app.getHttpHelperInstance().call(user, method, getUrl(itemId));
+
+    if (response.statusCode === HttpStatusCode.OK) {
```
Using a service will make this better, also, look at my point in fetchComments

Gustrb commented:
Ok so, I took some time to look at your code, and it is getting pretty good, congratulations 🚀 
But I have A few issues I found during the review.
Most of them are due to:
- Poor logging (you'll thank me later for being annoying with this, logging is our friend when everything stops working)
- No domaing splitting (you are grouping functions by "what they do" instead of by "what they are"), doing that will make the code easier to read and navigate
- Mixing up what is domain and what is infraestructure, please, avoid doing direct HTTP calls from your UI components, where you build your UI should not care about HTTP stuff such as the error that came back, or if the body is empty. When you start defering that to a service class, it will become clear there are a few issues with how everything is being handled, so the first issue is:
1. There are not a lot of validations on what we are sending the network (nothing on the typesystem): Since our HTTP calls go through a bridge, then through (possibliy) a proxy to get over to the Todoist backend, it might be really slow, so we must be pedantic and only allow http calls that are going to succeed.
2. There are no validations on the recieving end, everything is being typed as any and then being typed accordingly through a type assertion, this is *really* dangerous, since we have no idea when the api changes, and small things might start breaking slowly as the time goes by, and we will have 0 visibility into it. So please, use a library to validate your json schema from the response (and type it accordingly in the typesystem, so it is easier to mantain, and one needn't to have the API docs open to change simple stuff in the code)

Gustrb commented:
Also, dropping here a cool article about defensive programming, I think you'll benefit a lot from thinking like this: https://medium.com/@arumukherjee121/the-art-of-defensive-programming-62c6f22b2758

Since it is really annoying to update apps, and manage your deployed apps, we need to get the first version as stable as possible, and we do it, by programming defensively

jannatkhandev commented:
I really appreciate your time, the feedback and review really helps me grow as a developer.
As I am still learning and trying to improve, your review really made me understand industry standards.

I will make all the improvements over the weekend and will get ready by Monday addressing all the comments.

Yet again, thanks for your time and feedback.
Best,
Jannat :)

Gustrb commented:
Don't worry, I'm happy to help :)
I still haven't started doing actual testing, but after this next round of reviews we can start doing this.
You're doing a great job, keep going 🚀 

Also, don't worry about being quick to ship stuff, take your time, ponder on your code, experiment, change stuff, that is really important for your growth as a developer.
Have a good weekend!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/RocketChat/App_Todoist?pullRequest=1) <br/>All committers have signed the CLA.

jannatkhandev committed:
Add review changes

jannatkhandev commented:
Hello sir,

I deeply retrospected the code and made all requested changes, made the app more robust and code easy to follow.
Please take look.

(Also sorry for messing up the commits, I changed my machine [now I run VPS on a server and use it in VSCode via SSH])

jannatkhandev committed:
Update GitIgnore

Gustrb commented on TodoistApp.ts:
```diff
@@ -0,0 +1,139 @@
+    };
+
+    constructor(info: IAppInfo, logger: ILogger, accessors: IAppAccessors) {
+        super(info, logger, accessors);
```
why not just:

this.labelService = new LabelService(this);
...

jannatkhandev replied:
If I call the constructor here it goes into a cyclic loop of invocations as we use TodoistApp in the constructor of each service, hence.

Gustrb replied:
why cyclic loop? no service is instantiating the app class

I guess it could not work bc the services would be getting a partially uninitialized memory, but the pointer would still be the same, so I am not sure

Gustrb replied:
seems like there is no problem: https://www.typescriptlang.org/play/?#code/MYGwhgzhAEBC0G8BQ1XQA4CcCWA3MALgKbRgBc0AggNwprAD2AdhAZgK7AEOYAU5VAJSI6aNAQAW2CADow0ALylaYgL5JR0AOZEClXsORixmXe0xNok6XJVp16pKEgxKIsVjyESAIwqw7VEYWNk5uPkNNMWtZH0VoJiIAdzheGMFA6EcxHQJYA3djVFMCc0sYmR8ZXP0MzQcNYNZSeMSU2tpsADNofmrdfOEAQgUlMEjoiUwGFLboAFFMab4AciSibEwAExW61SA

Gustrb commented on src/modals/createTaskModal.ts:
```diff
@@ -0,0 +1,120 @@
+  };
+}
+
+async function createProjectSection(projectId?: string): Promise<LayoutBlock[]> {
```
why is this async?

Gustrb commented on src/modals/createTaskModal.ts:
```diff
@@ -0,0 +1,120 @@
+  return [projectIdInputBox];
+}
+
+async function createTaskDetailsSection(descriptionText?: string): Promise<LayoutBlock[]> {
```
why is this async?

Gustrb commented on src/services/SharedLabelService.ts:
```diff
@@ -0,0 +1,51 @@
+        throw new Error(`Failed to fetch labels: ${response.content}`);
+      }
+
+      const labels: ILabel[] = response.data;
```
you are not validating if response.data is not an array.
it might be null... or just a plain object

Gustrb commented on src/services/TaskService.ts:
```diff
@@ -0,0 +1,78 @@
+      if (response.statusCode !== 200) {
+        throw new Error(`Failed to fetch tasks: ${response.content}`);
+      }
+      const tasks: ITask[] = Array.isArray(response.data) ? response.data : [response.data];
```
I think you should be a little less optimistic here.
you cannot know what the type is. So what I would do is something like this:


```suggestion
      this.validate(response.data);

      ...

    private validate(tasks: unknown): void {
        if (tasks === null || tasks === undefined || !Array.isArray(tasks)) {
            throw new Error(`Invalid task structure: ${JSON.stringify(task)}`);
        }

        tasks.forEach((task) => {
            if (typeof task !== object || task === null) {
                throw new Error(`Invalid task structure: ${JSON.stringify(task)}`);
            }

            if (!('id' in task) || typeof task.id !== string) {
                throw new Error(`Invalid task structure: ${JSON.stringify(task)}`);
            }

            if (!('content' in task) || typeof task.content !== string) {
                throw new Error(`Invalid task structure: ${JSON.stringify(task)}`);
            }
        });
    }

```

Gustrb commented:
Looking nice, I just think your validations are a little to optimistic, besides that, the code is looking good, congrats!
